pubtools/exodus/hooks/pulp.py

```diff
@@ -0,0 +1,165 @@
+                raise outer
+
+            LOG.error("Unsuccessful response from exodus-gw: %s", body)
+            raise
+
+    @backoff.on_predicate(
+        wait_gen=backoff.expo,
+        predicate=lambda task: task["state"] not in ("FAILED", "COMPLETE"),
+        max_time=1800,
+    )
+    def _poll_commit_completion(self, commit):
+        """Issues request(s) to exodus-gw to get the task state, returning once
+        it either is FAILED or COMPLETED."""
+        task_url = urljoin(self._gw_url, commit["links"]["self"])
+
+        resp = self._do_request(method="GET", url=task_url)
+        task = resp.json()
+
+        LOG.info("Task state: %s", task["state"])
+
+        return task
+
+    @hookimpl
+    def pulp_repository_pre_publish(self, to_publish):
+        """Invoked before a Pulp repository is published.
+
+        This implementation adds to each config the --exodus-publish argument,
+        attaching the repository to an Exodus gateway publish.
+
+        Args:
+            to_publish (list[tuple[:class:`~pubtools.pulplib.Distributor`, dict]])
+                List of Distributor-config pairs intended for publishing.
+
+        Returns:
+            to_publish (list[tuple[:class:`~pubtools.pulplib.Distributor`, dict]])
+                List of potentially adjusted Distributor-config pairs to publish.
+        """
+
+        if not self._publish:
+            publish_url = os.path.join(self._gw_url, self._gw_env, "publish")
+            resp = self._do_request(method="POST", url=publish_url)
+            self._publish = resp.json()
+            LOG.debug("Created publish: %s", self._publish["id"])
+
+        publish_id = self._publish["id"]
+
+        for (_, config) in to_publish:
+            rsync_extra_args = config.get("rsync_extra_args", [])
+            rsync_extra_args.append("--exodus-publish=%s" % publish_id)
+            config["rsync_extra_args"] = rsync_extra_args
+
+        return to_publish
+
+    @hookimpl
+    def task_pulp_flush(self):
+        """Invoked during task execution after successful completion of all
+        Pulp writes and publishes.
+
+        This implementation commits the active Exodus gateway publish, making
+        the content visible on the target CDN environment.
+        """
+
+        if not self._publish:
+            LOG.info("No publish to commit.")
+            return
+
+        commit_url = urljoin(self._gw_url, self._publish["links"]["commit"])
+        resp = self._do_request(method="POST", url=commit_url)
+        commit = resp.json()
+
+        task = self._poll_commit_completion(commit)
+        LOG.info("Publish complete: %s", task)
+
+
+@hookimpl
+def task_start():
+    enable_vals = ["true", "t", "1", "yes", "y"]
+    enabled = os.getenv("EXODUS_ENABLED", "False").lower() in enable_vals
+
+    if enabled and not pm.is_registered(ExodusPulpHanlder()):
+        pm.register(ExodusPulpHanlder())
+
+
+@hookimpl
+def task_stop():
+    if pm.is_registered(ExodusPulpHanlder()):
```

Comment:
This shouldn't work for the same reason as mentioned in the task_start comment, this is just creating a new Handler each time, which will never be registered so this will always be false.

You should be able to put the task_stop hookimpl as an instance method on ExodusPulpHandler, could you try that if you haven't already?  Like this:

```python
 @hookimpl
 def task_stop(self):
     pm.unregister(self)
```

What's supposed to happen is that, when you pm.register(ExodusPulpHandler()), that register will find the task_stop instance method and you don't need any module-level task_stop at all.  I thought the pubtools docs had an example like that, but it seems not.